xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java

```diff
@@ -18,47 +17,72 @@
 
     private static final Logger LOG = LoggerFactory.getLogger(KrakenStreamingExchange.class);
-
     private static final String API_URI = "wss://ws.kraken.com";
+    private static final String API_AUTH_URI = "wss://ws-auth.kraken.com";
+    private static final String API_BETA_URI = "wss://beta-ws.kraken.com";
 
-    private final KrakenStreamingService streamingService;
+    private KrakenStreamingService streamingService, privateStreamingService;
     private KrakenStreamingMarketDataService streamingMarketDataService;
-    
+    private KrakenStreamingTradeService streamingTradeService;
+
     private final SynchronizedValueFactory<Long> nonceFactory = new CurrentTimeNonceFactory();
 
     public KrakenStreamingExchange() {
-        this.streamingService = new KrakenStreamingService(API_URI);
+    }
+
+    private static String pickUri(boolean isPrivate, Boolean useSandbox) {
+        if (useSandbox != null && useSandbox)
+            return API_BETA_URI;
+        else
+            return isPrivate ? API_AUTH_URI : API_URI;
     }
 
     @Override
     protected void initServices() {
         super.initServices();
-        streamingMarketDataService = new KrakenStreamingMarketDataService(streamingService);
+        Boolean useSanbox = (Boolean)exchangeSpecification.getExchangeSpecificParametersItem(USE_SANDBOX);
+
+        this.streamingService = new KrakenStreamingService(false, pickUri(false,useSanbox));
+        this.streamingMarketDataService = new KrakenStreamingMarketDataService(streamingService);
+
+        if (exchangeSpecification.getApiKey() != null) {
+            this.privateStreamingService = new KrakenStreamingService(true, pickUri(true,useSanbox));
+        }
+
+        KrakenAccountServiceRaw rawKrakenAcctService = (KrakenAccountServiceRaw) getAccountService();
+
+        streamingTradeService = new KrakenStreamingTradeService(privateStreamingService, rawKrakenAcctService);
     }
 
     @Override
     public Completable connect(ProductSubscription... args) {
+        if (privateStreamingService != null)
+            return privateStreamingService.connect().andThen(streamingService.connect());
+
         return streamingService.connect();
     }
 
     @Override
     public Completable disconnect() {
+        if (privateStreamingService != null)
+            return privateStreamingService.disconnect().andThen(streamingService.disconnect());
+
         return streamingService.disconnect();
     }
 
     @Override
     public boolean isAlive() {
-        return streamingService.isSocketOpen();
+        return streamingService.isSocketOpen() && (privateStreamingService == null || privateStreamingService.isSocketOpen());
     }
     
     @Override
     public Observable<Object> connectionSuccess() {
-        return streamingService.subscribeConnectionSuccess();
+        return streamingService.subscribeConnectionSuccess().mergeWith(privateStreamingService.subscribeConnectionSuccess());
```

Comment:
Hmmm. It seems a little unexpected that you'll get two connection successes. I wonder what people use this observable for?  With this possibility, you can't rely on it for state on its own; you need to check `isAlive()` as well, which should probably documented if we think this is the right behaviour.

Reply:
Not sure what to do about this.  Wait for both, then emit one.  kind of like debounce.

Reply:
I suppose that's probably the most predictable option, yes!

Reply:
558e6f61ed4922f3fea4625241a71f6a8fefac0d